### PR TITLE
R.I.P.  Mr. SFKeychainItemWrapper 

### PR DESCRIPTION
**Pull request by trooper2013**
This PR reworks how we perform crud operations with the Keychain. The follow up PR's will take care of keychain item accessibility changes. This should only change the code used to perform crud operations. The new code does not attempt to reorganize the keys differently. As such no migration is required. Hopefully we won't have to bring back SFKeychainItemWrapper from the dead :)

Does the following.

1.  Removes the usage of SFKeyChainItemWrapper (non-arc) class.
2. Adds a new Swift based KeychainHelper to the SalesforceSDKCommon.
3. Refactors code and some tests accordingly.
4. Adds a basic unit test.

TODO

- [x] Adding comprehensive Unit Tests for the keychain helper
- [x] Deprecate SFKeychainItemWrapper.
- [x] Test apps with diferent setups(passcode enabled etc..)
- [ ] Handle changing of accessibility attribute (Follow up PR)

### CODE REVIEW

File: libs/SalesforceSDKCommon/SalesforceSDKCommonTests/KeychainItemManagerTests.swift [hunk 1]
```diff
@@ -0,0 +1,103 @@
+//
+//  KeychaiinItemManagerTests.swift
```
Comment (wmathurin) [inline libs/SalesforceSDKCommon/SalesforceSDKCommonTests/KeychainItemManagerTests.swift:2]: Typo (double i)

**Review by wmathurin — COMMENTED** [review]
You should be able to simplify the podspec (see https://github.com/forcedotcom/SalesforceMobileSDK-iOS/blob/dev/SalesforceSDKCore.podspec#L36).

File: libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+internal class KeychainItemManager: NSObject {
+    private let secureStoreQueryable: KeychainItemQueryable
+    static let errorDomain = "com.salesforce.security.keychainException"
+    
+    /// Initializer for kSecClassGenericPassword with accessgroup. Will create a keychain item manager for kSecClassGenericPassword operations
+    init(service: String, account: String?) {
+        self.secureStoreQueryable = GenericPasswordItemQuery(service: service,
+                                                             account: account,
+                                                             accessGroup: nil)
```
Comment (AsceticMonk) [inline libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift:67]: Should this one simply call the initializer below?

File: libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+        
+        let additions: [String: Any] = [String(kSecMatchLimit): kSecMatchLimitOne,
+                                        String(kSecReturnAttributes): kCFBooleanTrue as Any,
+                                        String(kSecReturnData): kCFBooleanTrue as Any]
+        
+        var query = secureStoreQueryable.query
+        query.merge(additions) { (current, _) in current }
+        
+        var queryResult: CFTypeRef?
+        let  status = SecItemCopyMatching(query as CFDictionary, &queryResult)
```
Comment (AsceticMonk) [inline libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift:144]: nib: extra space between `let` and `status`

File: libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    }
+}
+
+
+@objc(SFSDKKeychainResult)
+public class KeychainResult: NSObject {
+    @objc public let success: Bool
+    @objc public var status: OSStatus
+    @objc public var data: Data?
+    @objc public var error: NSError?
```
Comment (AsceticMonk) [inline libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift:190]: Should `status`, `data`, and `error` be `let` instead? Do they mutate after initialization?

File: libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    /// - Parameters:
+    ///   - service: Identifier to use for keychain service key.
+    ///   - account: Identifier to use for keychain account key.
+    /// - Returns: KeychainResult
+    @objc public class func read(service: String, account: String?) -> KeychainResult {
+        let keychainManager = KeychainItemManager(service: service, account: account)
+        return keychainManager.getValue()
+    }
+    
+
```
Comment (AsceticMonk) [inline libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift:218]: nil: extra newline

File: libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+    
+    
+    /// Create an item in the keychain if not present.
+    /// - Parameters:
+    ///   - service: Identifier to use for keychain service key.
+    ///   - account: Identifier to use for keychain account key.
+    /// - Returns: KeychainResult
+    @objc public class func createIfNotPresent(service: String, account: String?) -> KeychainResult {
+        let keychainManager = KeychainItemManager(service: service, account: account)
+        var keychainResult =  keychainManager.getValue()
```
Comment (AsceticMonk) [inline libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift:226]: nib: extra space after `=`

File: libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+        if !keychainResult.success && keychainResult.status == errSecItemNotFound {
+            keychainResult =  keychainManager.addEmptyValue()
```
Comment (AsceticMonk) [inline libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift:228]: nib: extra space after `=`

File: libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+    
+    /// If an item is found remove it and then add an empty entry i.e. data is nil.
+    /// - Parameters:
+    ///   - service: Identifier to use for keychain service key.
+    ///   - account: Identifier to use for keychain account key.
+    /// - Returns: KeychainResult
+    @objc public class func reset(service: String, account: String?) -> KeychainResult {
+        let keychainManager = KeychainItemManager(service: service, account: account)
+        var keychainResult = keychainManager.getValue()
+        if keychainResult.success == true, keychainManager.removeValue().success {
```
Comment (AsceticMonk) [inline libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift:252]: `success` is already a boolean, right? Do we need this comparison to `true`?

File: libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+    
+    /// Remove an item from the keychain.
+    /// - Parameters:
+    ///   - service: Identifier to use for keychain service key.
+    ///   - account: Identifier to use for keychain account key.
+    /// - Returns: KeychainResult
+    @objc public class func remove(service: String, account: String?) -> KeychainResult {
+        let keychainManager = KeychainItemManager(service: service, account: account)
+        var keychainResult = keychainManager.getValue()
+        if keychainResult.success == true {
```
Comment (AsceticMonk) [inline libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift:267]: Same comment here about the comparison to boolean literal

Comment (trooper2013) [reply libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift:267]: verbose. Ill fix it.

File: libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift (hunk 4 above)
Comment (trooper2013) [reply libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift:190]: When i started these were. They don't need to be. Thanks I will rework this.

File: libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift (hunk 2 above)
Comment (trooper2013) [reply libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift:67]: actually we don't need the other initializer for now. I will remove it.

**Comment by trooper2013** [comment]
@bhariharan @wmathurin Ive tested all of the following along with @bbirman 

-  Sample apps (native) 
-  Passcode flows.
-  Auth handling before and after. 
-  App group enablement.
-  Upgrade scenarios.

It's all working as expected.

**Review by bhariharan — COMMENTED** [review]
Could you also resolve the merge conflicts and push a patch?

File: libs/SalesforceSDKCommon/SalesforceSDKCommonTests/SalesforceSDKCommonTests-Bridging-Header.h [hunk 10]
```diff
@@ -0,0 +1,4 @@
+//
```
[the remark this answers is not in the document]
Comment (trooper2013) [reply libs/SalesforceSDKCommon/SalesforceSDKCommonTests/SalesforceSDKCommonTests-Bridging-Header.h:1]: @bhariharan I can remove this file, Did not understand the previous comment "remove what?"  For some reason I cant add a reply to that comment :)

**Review by bhariharan — APPROVED** [review]

**Pull request merged**